core/schema/module.go

```diff
@@ -1206,9 +1301,31 @@ func (s *moduleSchema) createIDResolver(typeDef *core.TypeDef, dest *schemaView)
 				}
 				return resolver.FromID(s)
 			}
+		} else {
+			return func(a any) (any, error) {
+				id, ok := a.(string)
+				if !ok {
+					if _, ok := a.(map[string]any); ok {
+						return a, nil
+					}
+					return nil, fmt.Errorf("expected map %s, or string %sID, got %T", typeDef.AsObject.Name, typeDef.AsObject.Name, a)
+				}
+
+				typeName, value, err := resourceid.DecodeModule(id)
+				if err != nil {
+					return a, nil
+				}
+				if typeName != typeDef.AsObject.Name {
+					// if we hit this, the module is misbehaving - it's
+					// declared it returns a different module type than it's
+					// actually returned
+					return nil, fmt.Errorf("invalid type name %q, expected %q", typeName, typeDef.AsObject.Name)
+				}
+				return value, nil
```

Comment:
nit: Seems like this sanity check should happen in `DecodeModule` for consistency with `(ID).Decode`.

```suggestion
				return resourceid.DecodeModule(id, typeDef.AsObject.Name)
```

Reply:
Yeah good call (here and related comment below), made this update.

It also got rid of a case where we ignored the error from `resourceid.DecodeModule`, we just always return an error from that if received here now. AFAICT that's the desirable behavior and the tests pass, so think we're good to go.